crates/admin/src/cluster_controller/service.rs

```diff
@@ -236,32 +240,106 @@ impl ClusterControllerHandle {
             tx: self.command_tx.clone(),
         }
     }
 
-    pub async fn run(mut self) -> anyhow::Result<()> {
-        self.init_partition_table().await?;
+    fn cluster_controller_status(&self) -> ClusterControllerStatus {
+        let nodes_config = self.metadata.nodes_config_ref();
+        let maybe_active = nodes_config
+            .get_admin_nodes()
+            .filter(|node| {
+                self.observed_cluster_state
+                    .is_node_alive(node.current_generation)
+            })
+            .map(|node| node.current_generation.as_plain())
+            .sorted()
+            .next();
+
+        // A Cluster Controller is active if the node holds the smallest plain Node ID
+
+        let maybe_active = !maybe_active
+            .is_some_and(|active_node_id| active_node_id != self.metadata.my_node_id().as_plain());
+
+        if let Err(err) = nodes_config.find_node_by_id(self.metadata.my_node_id()) {
+            warn!("Shuting down cluster controller because of a node generation error {err}");
+            return ClusterControllerStatus::Shutdown;
+        }
+
+        match maybe_active {
+            true => ClusterControllerStatus::Active,
+            false => ClusterControllerStatus::Passive,
+        }
+    }
+
+    async fn passive(
+        &mut self,
+        config_watcher: &mut ConfigWatch,
+        cluster_state_watcher: &mut ClusterStateWatcher,
+    ) -> anyhow::Result<ClusterControllerStatus> {
+        let mut shutdown = std::pin::pin!(cancellation_watcher());
 
         let bifrost_admin = BifrostAdmin::new(
             &self.bifrost,
             &self.metadata_writer,
             &self.metadata_store_client,
         );
 
+        loop {
+            tokio::select! {
+                _ = self.heartbeat_interval.tick() => {
+                    // Ignore error if system is shutting down
+                    let _ = self.cluster_state_refresher.schedule_refresh();
+                },
+                Ok(cluster_state) = cluster_state_watcher.next_cluster_state() => {
+                    self.observed_cluster_state.update(&cluster_state);
+                }
+                Some(cmd) = self.command_rx.recv() => {
+                    // it is still safe to handle cluster commands as a passive CC
+                    self.on_cluster_cmd(cmd, bifrost_admin).await;
+                }
+                _ = config_watcher.changed() => {
+                    debug!("Updating the cluster controller settings.");
+                    let options = &self.configuration.live_load().admin;
+
+                    self.heartbeat_interval = Self::create_heartbeat_interval(options);
+                    (self.log_trim_interval, self.log_trim_threshold) = Self::create_log_trim_interval(options);
+                }
+                _ = &mut shutdown => {
+                    self.health_status.update(AdminStatus::Unknown);
+                    return Ok(ClusterControllerStatus::Shutdown);
+                }
```

Comment:
I don't believe changes in logs and/or partition table can trigger a reconfiguration of leadership of the CC, what do you think?